File: MaximNigmatulin/Task4/src/main/kotlin/spbu/system/lazyset/LazySet.kt
package spbu.system.lazyset

import spbu.models.StudentRecord
import spbu.system.ExamSystemInterface

class LazySet : ExamSystemInterface {
    private val head = Node(
        next = null,
        key = -1,
        item = StudentRecord(-1, -1)
    )

    override fun add(record: StudentRecord): Boolean {
        val key = record.hashCode()
        while (true) {
            var pred = head
            var curr = pred.next

            while (curr != null && curr.key < key) {
                pred = curr
                curr = curr.next
            }
            pred.lock()
            try {
                curr?.lock()
                try {
                    if (validate(pred, curr)) {
                        if (curr?.key == key) {
                            return false
                        }
                        val insert = Node(curr, key, record)
                        pred.next = insert
                        return true
                    }
                } finally {
                    curr?.unlock()
                }
            } finally {
                pred.unlock()
            }
        }
    }

    override fun remove(record: StudentRecord): Boolean {
        val key = record.hashCode()
        while (true) {
            var pred = head
            var curr = pred.next

            while (curr != null && curr.key < key) {
                pred = curr
                curr = curr.next
            }
            pred.lock()
            try {
                curr?.lock()
                try {
                    if (validate(pred, curr)) {
                        if (curr?.key != key) {
                            return false
                        }
                        curr.marked = true
                        pred.next = curr.next
                        return true
                    }
                } finally {
                    curr?.unlock()
                }
            } finally {
                pred.unlock()
            }
        }
    }

    override fun contains(record: StudentRecord): Boolean {
        val key = record.hashCode()
        var curr: Node<StudentRecord>? = head
        while (curr != null && curr.key < key) {
            curr = curr.next
        }
        return curr != null && !curr.marked && curr.key == key
    }

    override fun count(): Int {
        var curr: Node<StudentRecord>? = head
        var count = 0
        while (curr != null) {
            curr = curr.next
            curr?.let {
                if (!curr.marked)
                    count += 1
            }
        }
        return count
    }

    private fun <StudentRecord> validate(pred: Node<StudentRecord>, curr: Node<StudentRecord>?): Boolean {
        var node: Node<spbu.models.StudentRecord>? = this.head
        while (node != null && node.key <= pred.key) {
            if (node == pred) {
                return pred.next == curr
            }
            node = node.next
        }
        return false
    }
}

File: KarimiHurmatullah/Task4/ExamSystem/ExamSystem/Collections/LazySet.cs
namespace ExamSystem.Collections
{
    public class LazySet<T>
    {
        private readonly Node<T> tail = new(int.MaxValue);
        private readonly Node<T> head = new(int.MinValue);
        private int count;

        public LazySet()
        {
            head.Next = tail;
        }

        private static bool Validate(Node<T> pred, Node<T> curr)
        {
            return !pred.Marked && !curr.Marked && pred.Next == curr;
        }

        public void Add(T item)
        {
            int key = item.GetHashCode();
            while (true)
            {
                Node<T> pred = head;
                Node<T> curr = head.Next;
                while (curr.Key < key)
                {
                    pred = curr;
                    curr = curr.Next;
                }
                pred.Lock.Lock();
                try
                {
                    curr.Lock.Lock();
                    try
                    {
                        if (Validate(pred, curr))
                        {
                            if (curr.Key == key)
                            {
                                return;
                            }
                            else
                            {
                                Node<T> node = new Node<T>(item) { Next = curr };
                                pred.Next = node;
                                Interlocked.Increment(ref count);
                                return;
                            }
                        }
                    }
                    finally
                    {
                        curr.Lock.Unlock();
                    }
                }
                finally
                {
                    pred.Lock.Unlock();
                }
            }
        }

        public void Remove(T item)
        {
            int key = item.GetHashCode();
            while (true)
            {
                Node<T> pred = head;
                Node<T> curr = head.Next;
                while (curr.Key < key)
                {
                    pred = curr;
                    curr = curr.Next;
                }
                pred.Lock.Lock();
                try
                {
                    curr.Lock.Lock();
                    try
                    {
                        if (Validate(pred, curr))
                        {
                            if (curr.Key != key)
                            {
                                return;
                            }
                            else
                            {
                                curr.Marked = true;
                                pred.Next = curr.Next;
                                Interlocked.Decrement(ref count);
                                return;
                            }
                        }
                    }
                    finally
                    {
                        curr.Lock.Unlock();
                    }
                }
                finally
                {
                    pred.Lock.Unlock();
                }
            }
        }

        public bool Contains(T item)
        {
            int key = item.GetHashCode();
            Node<T> curr = head;
            while (curr.Key < key)
                curr = curr.Next;
            return curr.Key == key && !curr.Marked;
        }

        public int Count => count;
    }

}
